service-base/src/main/java/org/eclipse/hono/service/NewAbstractApplication.java

```diff
@@ -0,0 +1,262 @@
+/**
+ * Copyright (c) 2017 Bosch Software Innovations GmbH.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *    Bosch Software Innovations GmbH - initial creation
+ */
+
+package org.eclipse.hono.service;
+
+import java.util.*;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+import javax.annotation.PostConstruct;
+import javax.annotation.PreDestroy;
+
+import org.eclipse.hono.config.ServiceConfigProperties;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.ObjectFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+
+import io.vertx.core.CompositeFuture;
+import io.vertx.core.Future;
+import io.vertx.core.Handler;
+import io.vertx.core.Vertx;
+
+/**
+ * TODO: fix me.
+ *
+ *
+ * A base class for implementing Spring Boot applications.
+ * <p>
+ * This class supports the configuration of an {@code ObjectFactory} for creating instances of the service
+ * implementation that this application exposes.
+ */
+public class NewAbstractApplication {
+
+    /**
+     * A logger to be shared with subclasses.
+     */
+    protected final Logger log = LoggerFactory.getLogger(getClass());
+
+    private final Set<ObjectFactory<? extends AbstractServiceBase>> serviceFactories = new HashSet<>();
+
+    private Vertx vertx;
+
+    /**
+     * Sets the Vert.x instance to deploy the service to.
+     *
+     * @param vertx The vertx instance.
+     * @throws NullPointerException if vertx is {@code null}.
+     */
+    @Autowired
+    public final void setVertx(final Vertx vertx) {
+        this.vertx = Objects.requireNonNull(vertx);
+    }
+
+    /**
+     * Gets the Vert.x instance the service gets deployed to.
+     *
+     * @return The vertx instance.
+     */
+    protected Vertx getVertx() {
+        return vertx;
+    }
+
+    /**
+     * Adds multiple serviceFactories to this server.
+     *
+     * @param definedServiceFactories The serviceFactories.
+     */
+    @Autowired(required = false)
+    public final void addServiceFactories(final Collection<ObjectFactory<? extends AbstractServiceBase>> definedServiceFactories) {
```

Comment:
any reason why the param is of type `Collection` and the property is of type `Set`?